### ISSUE TITLE
Convert all the docs to // CONSOLE

### ISSUE THREAD

**Issue by nik9000** [elastic/elasticsearch#18160]
**Describe the feature**:
#18075 will merge soon, testing all the `// CONSOLE` snippets in our docs. But lots of our docs _aren't_ annotated with it! We should annotate more snippets so we test more of the docs.

Edit: `// CONSOLE` has replaced `// AUTOSENSE`. The sense application is named console in 5.0.0.

**Comment by nik9000** [comment]
Running `gradle listAutoSenseCandidates` will spit out a list of snippets that maybe should be converted to `// AUTOSENSE`. I imagine it has a very very high false positive rate. If someone is so included, they can run it and convert what they can.

If we can get the false positive rate down to 0 then we can fail the build if this finds any snippets. *if*.

**Comment by geekpete** [comment]
Hi nik9000, do you have any easy guide on how to assist with converting?

**Comment by nik9000** [comment]
The simplest thing to convert to `// AUTOSENSE` is `curl` commands because they are already complete commands.

1. Remove the `curl -X` part. Sense doesn't need it.
2. Remove the `localhost:9200/` part. Sense has the port and host configured.
3. Remove the `-d'{` and put the opening `{` on the next line. Sense likes it there and the `-d'` is implied.
4. Remove the closing `'`.
5. If the source tag at the top of the snippet is `[source,sh]` or something switch it to `[source,js]`.

Now that should just work with `// AUTOSENSE`. You can read `docs/README.md` for any of the extra test markup. Stuff like `// TESTRESPONSE` for matching the response and `// TEST[continued]` for continuing on from the last snippet.

You can test your changes by running `gradle :docs:check`. The first time you do it will take a while, just like any java build. The list of things it has to do before it can start and skips if they are already done is immense. From downloading dependencies to compiling Elasticsearch to transforming the `.asciidoc` files into tests.

There are places where we didn't write the whole API call but I think those are easier once you know the sense syntax.

**Comment by clintongormley** [comment]
* What went wrong:
    Execution failed for task ':docs:buildRestTests'.
    > Path shouldn't start with a '/': reference/search.asciidoc[20:27](js)// AUTOSENSE
      POST /twitter/tweet?routing=kimchy
      {
        "user": "kimchy",
        "postDate": "2009-11-15T14:12:12",
        "message": "trying out Elasticsearch"
      }


It's perfectly fine for the path to start with a `/`.  In fact, I prefer it because it looks like a real path

**Comment by clintongormley** [comment]
> Remove the curl -X part. Sense doesn't need it.
> Remove the localhost:9200/ part. Sense has the port and host configured.
> Remove the -d'{ and put the opening { on the next line. Sense likes it there and the -d' is implied.
> Remove the closing '.
> If the source tag at the top of the snippet is [source,sh] or something switch it to [source,js].

Or paste it into Sense, which should convert it to Sense syntax.  Then press Ctrl-I or Cmd-I to reformat.

**Comment by nik9000** [comment]
> It's perfectly fine for the path to start with a /. In fact, I prefer it because it looks like a real path

I didn't like their being two ways to do it and it looked like more than half of the docs did it sans-`/` so I picked that way. No one seemed to complain but maybe no one noticed?

If you really want to have the inconsistency back I can support that too.

> Or paste it into Sense, which should convert it to Sense syntax. Then press Ctrl-I or Cmd-I to reformat.

The great irony of this is that I never use sense. I tried this morning and, of course, it didn't like that I was running on Elasticsearch's master branch and I hadn't built it from source. I don't have the patience nor the skill with javascript to run that from master too, I think.

**Comment by clintongormley** [comment]
> If you really want to have the inconsistency back I can support that too.

If anything, I'd prefer the other way.  But I'm OK with the inconsistency

**Comment by nik9000** [comment]
@clintongormley I created #18184 for it.

**Comment by MaineC** [comment]
@nik9000 How deep do we want those tests to be? TESTRESPONSE seems to be present only in a couple of documents.

Concrete example: For the query dsl docs - should there be more than just the AUTOSENSE annotation to include them in the test run so parse failures are caught?

Another questions: Should we/ do we barf in case there's deprecated stuff in those snippets?

**Comment by nik9000** [comment]
> @nik9000 How deep do we want those tests to be? TESTRESPONSE seems to be present only in a couple of documents.

I don't know. Just trying to run the query at all is a huge improvement over what we have. As a first pass I'd just `// AUTOSENSE`ify all of them.

If we want to go deeper we can use the `// TEST[setup:name_of_data_set]` markup to reference a dataset defined in `docs/build.gradle` and have that run before the snippet. Then you can make a `// TESTRESPONSE`.

In my first pass I was just trying to convert what we had, inventing syntax along the way. I just tried to pick up all the `// AUTOSENSE` stuff and do sensible things with the snippets that looked like responses. I don't know what the final state should be.



> Another questions: Should we/ do we barf in case there's deprecated stuff in those snippets?

Yes we should. No we don't.

I don't know that Elasticsearch has a parameter to reject deprecated stuff. If not we should invent a parameter and make a big deal out of it because it seems really useful. And all the snippets should send that parameter by default. I think some snippets will talk about deprecated syntax and we can make `// TEST[deprecated]` or something to allow the snippet to send deprecated syntax.

**Comment by MaineC** [comment]
> I don't know. Just trying to run the query at all is a huge improvement over what we have. As a 
> first pass I'd just // AUTOSENSEify all of them.

That's what I'm currently doing for the query-dsl docs (went through them about half a year ago already when switching the parsing code to ParseField, the more complicated json snippets ended up as *fromJSON tests in the *QueryBuilder Java tests back then, so I should recognize most of the issues I run into while going along)


> I don't know that Elasticsearch has a parameter to reject deprecated stuff.

For parsing queries through ParseField there is an option called "index.query.parse.strict" that will fail to parse deprecated json parameters with an exception.

**Comment by nik9000** [comment]
> For parsing queries through ParseField there is an option called "index.query.parse.strict" that will fail to parse deprecated json parameters with an exception.

Cool! Is that at the index level? I'd love something at the request level for this though I haven't looked at how hard it'd be to implement. And it'd be great if it were global including stuff like detecting deprecated mappings and stuff. It is the kind of thing we could have a lively discussion with @clintongormley about, I think.

**Comment by MaineC** [comment]
#17512 is related to this setting.

**Comment by nik9000** [comment]
Lol, I should remember my own issues.

**Comment by nik9000** [comment]
@clintongormley, you mentioned to me that you planned to rewrite some areas of the docs soon and would get to this as part of the rewrite. Like #18356. Anyway, do you have a list of places you plan to rewrite or a list of places you know you won't rewrite?

**Comment by clintongormley** [comment]
> Anyway, do you have a list of places you plan to rewrite or a list of places you know you won't rewrite?

Mappings I won't rewrite (or only a little).  Nor cluster and index settings.  Analysis is in progress.  That's about it - other docs I'll redo as I have the time

**Comment by MaineC** [comment]
Am I the only one having trouble to run `gradle :docs:check` even on master at the moment?

Running into a 

    ==> Test Info: seed=FA5BECA4868DAF99; jvm=1; suite=1
    Suite: org.elasticsearch.smoketest.SmokeTestDocsIT
    ERROR   0.01s | SmokeTestDocsIT.initializationError <<< FAILURES!
       > Throwable #1: java.lang.IllegalArgumentException: Negative position
       > 	at sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:717)
       > 	at org.elasticsearch.test.rest.parser.RestTestSuiteParser.parse(RestTestSuiteParser.java:54)
       > 	at org.elasticsearch.test.rest.ESRestTestCase.collectTestCandidates(ESRestTestCase.java:171)
       > 	at org.elasticsearch.test.rest.ESRestTestCase.createParameters(ESRestTestCase.java:148)
       > 	at org.elasticsearch.smoketest.SmokeTestDocsIT.parameters(SmokeTestDocsIT.java:40)
       > 	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
    Completed [1/1] in 0.01s, 1 test, 1 error <<< FAILURES!

**Comment by nik9000** [comment]
Clint had that a while back and needed to `gradle docs:clean` though I
don't know what caused it.

**Comment by MaineC** [comment]
That did the trick for me. Thanks.

Another question: Is it possible to limit test execution to just one doc file? That might speed up finding typos a bit when converting doc files with several snippets which look easy enough to not require trying each one of them separately.

**Comment by nik9000** [comment]
> Is it possible to limit test execution to just one doc file

`-Dtests.method="*file_name*"` is what I've been doing.

**Comment by nik9000** [comment]
I just checked. We currently run 553 docs tests and we have about 970 snippets in the docs that `gradle docs:listConsoleCandidates` thinks should be converted. I'm fairly sure it is over counting though I don't know by how much. We have more than a third but less than half of the snippets tested at this point.

**Comment by nik9000** [comment]
Removing the `v5.0.0` label because this isn't going to make 5.0.0. There are ~730 unconverted snippets.

**Comment by geekpete** [comment]
If we went totally crazy on it, do you rekon we could smash through those snippets?

**Comment by nik9000** [comment]
I'm not sure really. I know a pile of the snippets come from aggregations and for those it really helps to have the right set of data. I did a bunch of the pipeline aggregations ones and made a little data set in `docs/build.gradle` that matched well and everything worked out well.

I reckon we could split them up and say like "you do the `_cat` snippets, you do the aggregation snippets, you do the query snippets, etc" and after a couple of days work for a few people in parallel we could knock them out.

I try to knock out a few every few days, but at that rate it is going to take a while.

**Comment by nik9000** [comment]
@tlrx Finished this in 643eb286dcb18b4f2b59c0f5b3a554ffd2b35ca2. Thanks everyone doing all this work! 🍰

**Comment by geekpete** [comment]
Amazing effort! Awesome.
